src/sync/probspecs.nim

```diff
@@ -101,59 +102,81 @@ proc getNameOfRemote(probSpecsDir: ProbSpecsDir;
+    let rootCommitRef = gitCheck(0, ["rev-list", "--max-parents=0", "HEAD"],
+                                 "the directory at the cached " &
+                                 "problem-specifications location is not a " &
+                                 &"git repository: '{probSpecsDir}'")
+
+    if rootCommitRef != "8ba81069dab8e96a53630f3e51446487b6ec9212\n":
+      showError("the git repo at the cached problem-specifications location " &
+                &"has an unexpected initial commit: '{probSpecsDir}'")
 
     # Exit if the working directory is not clean (allowing untracked files).
-    discard gitCheck(0, ["diff-index", "--quiet", "HEAD"], "the given " &
+    discard gitCheck(0, ["diff-index", "--quiet", "HEAD"], "the cached " &
                      "problem-specifications working directory is not clean: " &
                      &"'{probSpecsDir}'")
 
-    # Find the name of the remote that points to upstream. Don't assume the
-    # remote is called 'upstream'.
-    # Exit if the repo has no remote that points to upstream.
-    const upstreamHost = "github.com"
-    const upstreamLocation = "exercism/problem-specifications"
-    let remoteName = getNameOfRemote(probSpecsDir, upstreamHost, upstreamLocation)
+    if conf.action.offline:
+      # Don't checkout `main` when `--offline` was passed, because:
+      # 1. The current tests of the configlet executable require that
+      #    `configlet sync --offline` does not alter the state of the cached
+      #    prob-specs repo.
+      # 2. It allows a power user to manually checkout an older commit in the
+      #    cached repo, which can help with making some atomic commits.
+      discard gitCheck(0, ["merge-base", "--is-ancestor", "HEAD", mainBranchName],
+                      &"the checked-out commit is not on the '{mainBranchName}'" &
+                      &"branch: '{probSpecsDir}'")
+    else:
+      # Checkout `main`. Like the above, don't allow a power user to sync from a
+      # prob-specs state that hasn't been merged to upstream `main`.
+      discard gitCheck(0, ["checkout", mainBranchName],
+                       &"failed to checkout '{mainBranchName}' in the cached " &
+                       &"problem-specifications directory: '{probSpecsDir}'")
+
+      # Find the name of the remote that points to exercism/problem-specifications.
+      # Don't assume the remote is named 'origin'.
+      # Exit if the repo has no such remote.
+      const upstreamHost = "github.com"
+      const upstreamLocation = "exercism/problem-specifications"
+      let remoteName = getNameOfRemote(probSpecsDir, upstreamHost, upstreamLocation)
 
-    if not conf.action.offline:
-      # For now, just exit with an error if the HEAD is not up-to-date with
-      # upstream, even if it's possible to do a fast-forward merge.
+      # `fetch` and `merge` separately, for better error messages.
+      logNormal(&"Running 'git pull' in cached problem-specifications dir...")
```

Comment:
Maybe have this be more about what is happening functionally?

```suggestion
      logNormal(&"Updating problem-specifications data...")
```

Or something like that.

Reply:
Pushed 9ed7f204e026 with extra "cached", like the help message. Let me know if you hate it.